### PR TITLE
Optimize fusion gru

### PR DESCRIPTION
**Pull request by tensor-tang**
add seq mode and refine batch mode.

### CODE REVIEW

File: paddle/fluid/operators/fusion_gru_op.cc [hunk 1]
```diff
@@ -13,16 +13,15 @@ See the License for the specific language governing permissions and
...
-#include "paddle/fluid/operators/math/detail/gru_cpu_kernel.h"
-#include "paddle/fluid/operators/math/detail/gru_kernel.h"
+#include "paddle/fluid/operators/math/cpu_vec.h"
 #include "paddle/fluid/operators/math/fc_compute.h"
-#include "paddle/fluid/operators/math/gru_compute.h"
-#include "paddle/fluid/operators/math/math_function.h"
 #include "paddle/fluid/operators/math/sequence2batch.h"
+#include "paddle/fluid/platform/cpu_info.h"
+
+DEFINE_bool(gru_use_seq, true, "Use sequence mode");
```
Comment (Superjomn) [inline paddle/fluid/operators/fusion_gru_op.cc]: Can this flag be replaced with an Op attribute?

There might be two gru, one use sequence mode, and another use batch mode.

File: paddle/fluid/operators/fusion_gru_op.cc [hunk 2]
```diff
@@ -115,22 +118,26 @@ void FusionGRUOpMaker::Make() {
...
+            "(LoDTensor) the result after X * WeightX (size is T x 3D)"
             " or batched_X (size is T x M), this will be automatically chosen,"
             " where T is the total time steps in this mini-batch,"
             " D is the hidden size, M is the dim size of x input.")
       .AsIntermediate();
-  AddOutput("BatchedGate", "(LoDTensor) Same as GRUOp").AsIntermediate();
-  AddOutput("BatchResetHiddenPrev", "(LoDTensor) (T x 3D) Same as GRUOp.")
-      .AsIntermediate();
-  AddOutput("BatchedHidden", "(LoDTensor) (T X D) Same as GRUOp.")
+  AddOutput("BatchedInput", "(LoDTensor) (T x 3D)").AsIntermediate();
```
Comment (Superjomn) [inline paddle/fluid/operators/fusion_gru_op.cc]: Some description here, the `BatchedInput` as Output seems a little weried.

File: paddle/fluid/operators/fusion_gru_op.cc [hunk 3]
```diff
@@ -153,172 +160,261 @@ more details can refer to GRU op.
...
-  dst->mutable_data<T>(src.dims(), ctx.GetPlace());
-  row_shuffle(ctx, src, index_lod, dst, indexed_src);
-}
-
-template <typename DeviceContext, typename T>
+template <typename T>
 class FusionGRUKernel : public framework::OpKernel<T> {
  public:
   void Compute(const framework::ExecutionContext& ctx) const override {
+    if (FLAGS_gru_use_seq) {
```
Comment (Superjomn) [inline paddle/fluid/operators/fusion_gru_op.cc]: The flag seems can be replaced with an Attr

File: paddle/fluid/operators/math/cpu_vec.h [hunk 4]
```diff
@@ -132,6 +132,121 @@ inline void vec_scal<float, platform::jit::avx512_common>(const int n,
   vec_scal<float, platform::jit::avx2>(n, a, x, y);
 }
 
+template <typename T, platform::jit::cpu_isa_t isa = platform::jit::isa_any>
+inline void vec_bias_sub(const int n, const T a, const T* x, T* y) {
```
Comment (Superjomn) [inline paddle/fluid/operators/math/cpu_vec.h:136]: Can `vec_bias_sub` replaced with a more general name, it is a general math equation `Y = a - X` ? 
same with the below ones

File: paddle/fluid/operators/fusion_gru_op.cc [hunk 5]
```diff
@@ -153,172 +160,261 @@ more details can refer to GRU op.
...
+    auto wh_dims = wh->dims();  // D x 3D
+    const int N = x_lod[0].size() - 1;
+    const int total_T = x_dims[0];
+    const int M = x_dims[1];
+    const int D3 = wh_dims[1];
+    const int D = wh_dims[0];
+    const int D2 = D * 2;
 
+    const T* x_data = x->data<T>();
+    const T* h0_data = h0 ? h0->data<T>() : NULL;
```
Comment (Superjomn) [inline paddle/fluid/operators/fusion_gru_op.cc]: Replace all NULL with `nullptr`, NULL is a `#define`d integer zero, while `nullptr` is a pointer value with type `void*`, more specific and safe.

https://www.quora.com/Whats-the-difference-between-NULL-and-nullptr-in-C++#

File: paddle/fluid/operators/fusion_gru_op.cc (hunk 1 above)
Comment (tensor-tang) [reply paddle/fluid/operators/fusion_gru_op.cc]: Sure, of course.
But flag should be more convenient.

File: paddle/fluid/operators/fusion_gru_op.cc (hunk 2 above)
Comment (tensor-tang) [reply paddle/fluid/operators/fusion_gru_op.cc]: OK, sure.

File: paddle/fluid/operators/fusion_gru_op.cc (hunk 3 above)
Comment (tensor-tang) [reply paddle/fluid/operators/fusion_gru_op.cc]: Yes.

File: paddle/fluid/operators/fusion_gru_op.cc (hunk 5 above)
Comment (tensor-tang) [reply paddle/fluid/operators/fusion_gru_op.cc]: OK, thx.

File: paddle/fluid/operators/math/cpu_vec.h (hunk 4 above)
Comment (tensor-tang) [reply paddle/fluid/operators/math/cpu_vec.h:136]: I though this name should be clear. 

Vector operation: one bias sub one vector then get other vector, corresponding to `vec_sub_bias` meaning `Y = X - a`.

File: paddle/fluid/operators/fusion_gru_op.cc (hunk 1 above)
Comment (Superjomn) [reply paddle/fluid/operators/fusion_gru_op.cc]: The Op's attribute can be fully controlled by IR or analysis, some attempt can be carry out on determine this attribute before the inference phase begin, and try to get the best performance.

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**